TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.PerformTarget.cs

```diff
@@ -0,0 +1,95 @@
+﻿using System.Collections.Generic;
+using TypeCobol.Compiler.CodeElements;
+using TypeCobol.Compiler.Nodes;
+using SectionNode = TypeCobol.Compiler.Nodes.Section;
+
+namespace TypeCobol.Analysis.Cfg
+{
+    public partial class ControlFlowGraphBuilder<D>
+    {
+        /// <summary>
+        /// PerfomrTarget class to capture target sentences and procedures from a PERFORM instruction.
+        /// </summary>
+        private class PerformTarget
+        {
+            /// <summary>
+            /// All target sentences
+            /// </summary>
+            internal List<Sentence> Sentences { get; private set; }
+            /// <summary>
+            /// All target procedures.
+            /// </summary>
+            internal List<Procedure> Procedures { get; private set; }
+
+            public PerformTarget(List<Sentence> sentences, List<Procedure> procedures)
+            {
+                this.Sentences = sentences;
+                this.Procedures = procedures;
+            }
+        }
+
+        /// <summary>
+        /// Cache of PerformTarget instances already calculated for a Perform procedure 
+        /// </summary>
+        private Dictionary<PerformProcedure, PerformTarget> _performTargetCache;
+        private PerformTarget ComputePerformTarget(PerformProcedure p, SectionNode sectionNode)
+        {
+            if (_performTargetCache == null)
+                _performTargetCache = new Dictionary<PerformProcedure, PerformTarget>();
+            if (_performTargetCache.TryGetValue(p, out var target)) {
+                return target;
+            }
+
+            SymbolReference procedureReference = p.CodeElement.Procedure;
+            SymbolReference throughProcedureReference = p.CodeElement.ThroughProcedure;
+
+            Node procedureNode = ResolveProcedure(p, sectionNode, procedureReference);
+            if (procedureNode == null)
+                return null;
```

Comment:
This remark goes for all the `return null;` instructions in this method. We should actually store the `null` value inside the dictionary to prevent attempting resolving the target multiple times.
Even better: we should create a special instance `Unresolved` with empty sentences and procedures to use whenever a `PerformProcedure` cannot have its target properly resolved. This way the method would always return a non-null value.

Reply:
I would say this an abnormal situation it should never happened, let's take it in account with an assert.

Reply:
In fact calling an inexisting procedure is a semantic error, so no need to assert here, and simply return null here without storing such error value.

Reply:
It is not a good practice to store null values for semantic error reason.